accepted/exception-handling.md

```diff
@@ -0,0 +1,240 @@
+
+* No dependence on `libunwind`. This library provides a C API for
+  determining and unwinding call chains in ELF programs. Nonetheless,
+  we do not want to bring it into our trusted computing base (TCB),
+  because we have found its implementations to be buggy in the past,
+  and such bugs can easily compromise the integrity of Wasmtime (some
+  past `libunwind` bugs:
+  [#2808](https://github.com/bytecodealliance/wasmtime/issues/2808),
+  [#3256](https://github.com/bytecodealliance/wasmtime/issues/3256),
+  [#7997](https://github.com/bytecodealliance/wasmtime/issues/7997)).
+
+* Fast unwinding strategy. We require unwinding to be fast enough to
+  support [the guest
+  profiler](https://docs.wasmtime.dev/examples-profiling-guest.html). Thus,
+  it is important that the chosen implementation strategy does not box
+  ourselves into a corner. Though, we consider it acceptable if the
+  minimal viable product (MVP) of the implementation does not achieve
+  peak performance for unwinding as long as there is a viable path to
+  improving the performance.
+
+* Compatibility with standard formats. Regardless of unwinding
+  strategy, Wasmtime must continue to work with system profilers which
+  suport DWARF or frame pointers, e.g. `perf`. Furthermore, in
+  Cranelift `cg_clif` should be extended with capabilities to leverage
+  the unwind info support to emit appropriate DWARF, Windows
+  Structured Exception Handling, and so forth. Though, we believe a
+  MVP capable of recovering the `vmctx` register should suffice for
+  critical use-cases.
+
+# Non-requirements
+[non-requirements]: #non-requirements
+
+* No support for the legacy exception handling
+  revision. Justification: the legacy revision is being phased out.
+
+* No support for unwinding across host frames. Justification:
+  unwinding across host frames would require implementing a full DWARF
+  unwinder or equivalent. Note: this just not preclude throwing and
+  catching exceptions when there is a host frame present in the call
+  stack, as we can catch and rethrow exceptions in trampolines on the
+  boundary.
+
+# Proposal sketch
+[proposal]: #proposal
+
+This section outlines the technical details of the proposal for
+implementing support for exception handling in Wasmtime and Cranelift.
+
+## CLIF extensions
+[clif-semantics]: #clif-semantics
+
+We propose two extensions to CLIF.
+
+* A new block type `catch` which is a landing pad for exceptions. This
+  type of block will have exactly one parameter which will be a
+  pointer-sized integer (morally the exception value), e.g. in CLIF
+  syntax
+```clif
+catch block123(v456: i64):
+  ...
+```
+  We disallow regular control flow edges to `catch` blocks. Meaning
+  that neither `jump`, `brif`, or `br_table` are allowed to branch
+  into a `catch` block. Instead, the control flow edges to `catch`
+  must come via a `try_call` instruction.
+
+* A new call instruction `try_call <ok_label>, <exception_label>`,
```

Comment:
One part perhaps worth pointing out here is that DWARF supports multiple unwind locations per `try_call`, so `<exception_label>` may want to actually be a list of labels. I believe that @bjorn3's initial work for rustc_codegen_cranelift modeled this with a `JumpTable` where the "default label" was the `ok_label` and everything else was an unwind location.